tests/tests-hz38/src/main/java/com/hazelcast/simulator/tests/cardinalityestimator/CardinalityEstimatorTest.java

```diff
@@ -0,0 +1,116 @@
+
+import com.hazelcast.cardinality.CardinalityEstimator;
+import com.hazelcast.core.IAtomicLong;
+import com.hazelcast.simulator.test.AbstractTest;
+import com.hazelcast.simulator.test.BaseThreadState;
+import com.hazelcast.simulator.test.annotations.Setup;
+import com.hazelcast.simulator.test.annotations.Teardown;
+import com.hazelcast.simulator.test.annotations.TimeStep;
+import com.hazelcast.simulator.test.annotations.Verify;
+
+import static org.junit.Assert.assertTrue;
+
+/**
+ * The CardinalityEstimatorTest can be used to verify the Cardinality Estimation behavior. This test asserts the
+ * estimation based on the following flow
+ * 1. Test will add a configurable number (i.e batchSize) of elements in Cardinality Estimator
+ * 2. Addition  of elements can have duplicates key based on duplicateKeysPercentage configured
+ * 3. In assertion, test will verify that the element presents in Cardinality Estimator is not deviated more that
+ * tolerance configured
+ */
+
+public class CardinalityEstimatorTest extends AbstractTest {
+
+    // properties‚
+    public long batchSize = 1000000;
+    public double tolerancePercentage = 0.1;
+    public double duplicateKeysPercentage = 0.1;
+
+
+    private IAtomicLong elementCounter;
+    private IAtomicLong rangeSelectorCounter;
+    private CardinalityEstimator cardinalityEstimator;
+
+    @Setup
+    public void setup() {
+        cardinalityEstimator = targetInstance.getCardinalityEstimator(name);
+        elementCounter = targetInstance.getAtomicLong(name);
+        rangeSelectorCounter = targetInstance.getAtomicLong("rangeSelector");
+        rangeSelectorCounter.set(100L);
+    }
+
+    @TimeStep
+    public void TimeStep(BaseThreadState state) {
+        Range batchRange = getRange(state);
+        logger.info("Running range : " + batchRange.toString());
+        for (long i = batchRange.startIndex; i < batchRange.endIndex; i++) {
+            cardinalityEstimator.add(i);
+        }
+    }
+
+    @Verify
+    public void verify() {
+        int tolerance = (int) (elementCounter.get() * tolerancePercentage);
+        logger.info("Actual Keys added : " + elementCounter.get());
+        logger.info("Acceptable Count(i.e. tolerance) : " + (elementCounter.get() - tolerance));
+        logger.info("Cardinality Estimation : " + cardinalityEstimator.estimate());
+        assertTrue(Math.abs(cardinalityEstimator.estimate() - elementCounter.get()) < tolerance);
+    }
+
+    private Range getRange(BaseThreadState state) {
+        if (rangeSelectorCounter.getAndIncrement() % (duplicateKeysPercentage * 100) != 0) {
+            return nextRange();
+        } else {
+            return randomRange(state);
+        }
+    }
+
+    private Range nextRange() {
+        long currentCounterValue = elementCounter.getAndAdd(batchSize);
+        return new Range(currentCounterValue, currentCounterValue + batchSize - 1);
+    }
+
+    private Range randomRange(BaseThreadState state) {
```

Comment:
I believe the following comment which got lost due to your changes is still valid.

```
the calculation of randomValue is problematic.

Lets assume lowerRange = 0. ElementCounter = 10000, batchSize=1000, so upperRange = 10.0000-1000 ==> 9000.

For random.nextLong we return 1.000.000.0000.

Now randomValue = 0 + (1.0000.0000.000 * (9000-0)) ===> 9.000.000.000.000.

Now the final range will be 8.999.999.999.000 to 9000.000.000.000. This range is not a range of duplicates since the range should be between 0..10000.

Please check your code before making a pr instead of other people needing to debug your code.
```